core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -1822,6 +1823,35 @@ public static RelDataType extractLastNFields(RelDataTypeFactory typeFactory,
         type.getFieldList().subList(fieldsCnt - numToKeep, fieldsCnt));
   }
 
+  /**
+   * Returns whether the decimal value can be represented without information loss
+   * using the specified type.
+   * For example, 1111.11
+   * - cannot be represented exactly using DECIMAL(3, 1) since it overflows.
+   * - cannot be represented exactly using DECIMAL(6, 3) since it overflows.
+   * - cannot be represented exactly using DECIMAL(6, 1) since it requires rounding.
+   * - can be represented exactly using DECIMAL(6, 2)
+   *
+   * @param value  A decimal value
+   * @param toType A DECIMAL type.
+   * @return whether the value is valid for the type
+   */
+  public static boolean canBeRepresentedExactly(@Nullable BigDecimal value, RelDataType toType) {
+    assert toType.getSqlTypeName() == SqlTypeName.DECIMAL;
+    if (value == null) {
+      return true;
+    }
+    value = value.stripTrailingZeros();
+    if (value.scale() < 0) {
```

Comment:
it would be great also to have a test with negative scale

Reply:
Frankly negative scales are something I don't understand well.
Many SQL dialects do not accept negative scales, but Calcite allegedly does, although there aren't many tests about that.
Maybe we can file a separate issue about testing negative scales in Calcite, I expect it may uncover quite a few bugs.

Reply:
I also think that whether negative scales are accepted or not should be a property of the type system.

Reply:
I haven't checked it however if you think that it is large area to work with, then separate issue could be probably also ok

Reply:
Yes, I think that many places in Calcite assume implicitly that decimal scales cannot be negative.

Reply:
what do we do about this PR in particular?

Reply:
as you mentioned if there is a separate issue for negative scale is required, could you please file one in Calcite jira?
All the rest is ok to me, thanks